### PR TITLE
:construction: Add type checking using `mypy`

### PR DESCRIPTION
**Pull request by Jiaqi-Lv**
Add type checking using `mypy`, a static type checker for Python.

This PR involves adding type hints and checking the correctness of existing type hints in the code base.
A GitHub action has been created to run `mypy` after push and pull requests.
Currently, only `tiatoolbox/__init__.py` has been checked to work without errors with `mypy`.

### CODE REVIEW

File: tiatoolbox/__init__.py [hunk 1]
```diff
@@ -88,8 +91,10 @@ def read_registry_files(path_to_registry: str | Path) -> dict | str:
 
 
     """
+    if isinstance(path_to_registry, Path):
```
Comment (shaneahmed) [inline tiatoolbox/__init__.py:94]:
```suggestion
```

You do not need an if statement here. "path_to_registry = str(path_to_registry)" should be enough.

File: tiatoolbox/__init__.py [hunk 2]
```diff
@@ -101,14 +106,16 @@ def read_registry_files(path_to_registry: str | Path) -> dict | str:
...
     spec = importlib.util.spec_from_file_location(name, module_location)
+    if spec is None or spec.loader is None:
+        raise ModuleNotFoundError(name=name, path=str(module_location))
     loader = importlib.util.LazyLoader(spec.loader)
     spec.loader = loader
     module = importlib.util.module_from_spec(spec)
     sys.modules[name] = module
     loader.exec_module(module)
-    return module
+    return {name: module}
```
Comment (shaneahmed) [inline tiatoolbox/__init__.py:118]: What's the benefit of returning this as a dictionary?

File: tiatoolbox/__init__.py (hunk 1 above)
Comment (Jiaqi-Lv) [reply tiatoolbox/__init__.py:94]: Yes `Path` here causes mypy to fail. It seems that you cannot use `/` on `Path` objects, `/` only works for `PurePath` objects.

File: tiatoolbox/__init__.py (hunk 2 above)
Comment (Jiaqi-Lv) [reply tiatoolbox/__init__.py:118]: The return type for this function used to be `sys.modules`, but `sys.modules` is a variable so we can't use it as type. I looked up the type of `sys.modules`, it was `dict[str, ModuleType]` so I changed the return type to this, don't think there is any benefit of returning a dictionary.

Alternatively we change the return type of this function to `ModuleType` and keep everything else the same.

File: tiatoolbox/__init__.py [hunk 3]
```diff
@@ -71,8 +73,9 @@ def filter(self: DuplicateFilter, record: LogRecord) -> bool:
 
 
 # runtime context parameters
+rcParam: dict[str, str | Path | dict]  # noqa: N816
 rcParam = {  # noqa: N816
-    "TIATOOLBOX_HOME": Path.home() / ".tiatoolbox",
+    "TIATOOLBOX_HOME": Path.home() / Path(".tiatoolbox"),
```
Comment (shaneahmed) [inline tiatoolbox/__init__.py:78]:
```suggestion
    "TIATOOLBOX_HOME": Path.home() / ".tiatoolbox",
```

File: tiatoolbox/__init__.py [hunk 4]
```diff
@@ -88,6 +91,7 @@ def read_registry_files(path_to_registry: str | Path) -> dict | str:
 
 
     """
+    path_to_registry = str(path_to_registry)
```
Comment (shaneahmed) [inline tiatoolbox/__init__.py:94]:
```suggestion
    path_to_registry = str(path_to_registry)   # To pass tests with Python 3.8
```

**Review by shaneahmed — APPROVED** [review]

**Pull request merged**